File: test/CSharp/CodeCracker.Test/Performance/StringBuilderInLoopTests.cs
using CodeCracker.CSharp.Usage;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Testing;
using System.Threading.Tasks;
using Xunit;

namespace CodeCracker.Test.CSharp.Performance
{
    public class StringBuilderInLoopTests : CodeFixVerifier<StringBuilderInLoopAnalyzer, StringBuilderInLoopCodeFixProvider>
    {

        [Fact]
        public async Task WhileWithoutAddAssignmentExpressionDoesNotCreateDiagnostic()
        {
            const string source = @"
    namespace ConsoleApplication1
    {
        class TypeName
        {
            public void Foo()
            {
                while (DateTime.Now.Second % 2 == 0)
                {
                    Method();
                }
            }
            public void Method() { }
        }
    }";
            await VerifyCSharpHasNoDiagnosticsAsync(source);
        }

        [Fact]
        public async Task WhileWithoutStringConcatDoesNotCreateDiagnostic()
        {
            var source = @"
                var a = 0;
                while (a < 10)
                {
                    a += 1;
                }".WrapInCSharpMethod();
            await VerifyCSharpHasNoDiagnosticsAsync(source);
        }

        [Fact]
        public async Task WhileWithStringConcatOnLocalVariableCreatesDiagnostic()
        {
            var source = @"
                var myString = """";
                while (DateTime.Now.Second % 2 == 0)
                {
                    myString += """";
                }".WrapInCSharpMethod();
            var expected = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(14, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "myString"));
            await VerifyCSharpDiagnosticAsync(source, expected);
        }

        [Fact]
        public async Task WhileWithStringConcatOnFieldVariableCreatesDiagnostic()
        {
            const string source = @"
    namespace ConsoleApplication1
    {
        class TypeName
        {
            private string myString = """";
            public void Foo()
            {
                while (DateTime.Now.Second % 2 == 0)
                {
                    myString += """";
                }
            }
        }
    }";
            var expected = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(11, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "myString"));
            await VerifyCSharpDiagnosticAsync(source, expected);
        }


        [Fact]
        public async Task WhileWithStringConcatOnPropertyVariableCreatesDiagnostic()
        {
            const string source = @"
    namespace ConsoleApplication1
    {
        class TypeName
        {
            public string MyString { get; set; } = """";
            public void Foo()
            {
                while (DateTime.Now.Second % 2 == 0)
                {
                    MyString += """";
                }
            }
        }
    }";
            var expected = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(11, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "MyString"));
            await VerifyCSharpDiagnosticAsync(source, expected);
        }

        [Fact]
        public async Task WhileWithStringConcatWithSeveralConcatsOnDifferentVarsCreatesSeveralDiagnostics()
        {
            var source = @"
                var myString1 = """";
                var myString2 = """";
                while (DateTime.Now.Second % 2 == 0)
                {
                    myString1 += """";
                    myString2 += """";
                }".WrapInCSharpMethod();
            var expected1 = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(15, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "myString1"));
            var expected2 = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(16, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "myString2"));
            await VerifyCSharpDiagnosticAsync(source, expected1, expected2);
        }

        [Fact]
        public async Task WhileWithStringConcatWithSimpleAssignmentCreatesDiagnostic()
        {
            var source = @"
                var myString = """";
                while (DateTime.Now.Second % 2 == 0)
                {
                    myString = myString + """";
                }".WrapInCSharpMethod();
            var expected = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(14, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "myString"));
            await VerifyCSharpDiagnosticAsync(source, expected);
        }

        [Fact]
        public async Task WhileWithStringConcatWithSimpleAssignmentOnDifferentVarDoesNotCreateDiagnostic()
        {
            var source = @"
                var myString = """";
                var otherString = """";
                while (DateTime.Now.Second % 2 == 0)
                {
                    myString = otherString + """";
                }".WrapInCSharpMethod();
            await VerifyCSharpHasNoDiagnosticsAsync(source);
        }

        [Fact]
        public async Task FixesAddAssignmentInWhile()
        {
            var source = @"
                var myString = """";
                while (DateTime.Now.Second % 2 == 0)
                {
                    myString += ""a"";
                }".WrapInCSharpMethod();
            var fixtest = @"
                var myString = """";
                var builder = new System.Text.StringBuilder();
                builder.Append(myString);
                while (DateTime.Now.Second % 2 == 0)
                {
                    builder.Append(""a"");
                }
                myString = builder.ToString();".WrapInCSharpMethod();
            await VerifyCSharpFixAsync(source, fixtest);
        }

        [Fact]
        public async Task FixesAddAssignmentInWhileWithoutBlock()
        {
            var source = @"
                var myString = """";
                while (DateTime.Now.Second % 2 == 0)
                    myString += ""a"";".WrapInCSharpMethod();
            var fixtest = @"
                var myString = """";
                var builder = new System.Text.StringBuilder();
                builder.Append(myString);
                while (DateTime.Now.Second % 2 == 0)
                    builder.Append(""a"");
                myString = builder.ToString();".WrapInCSharpMethod();
            await VerifyCSharpFixAsync(source, fixtest);
        }

        [Fact]
        public async Task FixesAddAssignmentInWhileWithSystemTextInContext()
        {
            const string source = @"
    using System;
    using System.Text;
    namespace ConsoleApplication1
    {
        class TypeName
        {
            public void Foo()
            {
                var myString = """";
                while (DateTime.Now.Second % 2 == 0)
                {
                    myString += ""a"";
                }
            }
        }
    }";
            const string fixtest = @"
    using System;
    using System.Text;
    namespace ConsoleApplication1
    {
        class TypeName
        {
            public void Foo()
            {
                var myString = """";
                var builder = new StringBuilder();
                builder.Append(myString);
                while (DateTime.Now.Second % 2 == 0)
                {
                    builder.Append(""a"");
                }
                myString = builder.ToString();
            }
        }
    }";
            await VerifyCSharpFixAsync(source, fixtest, 0);
        }

        [Fact]
        public async Task FixesSimpleAssignmentInWhile()
        {
            var source = @"
                var myString = """";
                //comment 3
                while (DateTime.Now.Second % 2 == 0)
                {
                    //comment 1
                    myString = myString + ""a"";//comment 2
                }//comment 4".WrapInCSharpMethod();
            var fixtest = @"
                var myString = """";
                var builder = new System.Text.StringBuilder();
                builder.Append(myString);
                //comment 3
                while (DateTime.Now.Second % 2 == 0)
                {
                    //comment 1
                    builder.Append(""a"");//comment 2
                }//comment 4
                myString = builder.ToString();".WrapInCSharpMethod();
            await VerifyCSharpFixAsync(source, fixtest);
        }

        [Fact]
        public async Task FixesAddAssignmentWhenThereAre2WhilesOnBlock()
        {
            var source = @"
                var myString = """";
                while (DateTime.Now.Second % 2 == 1)
                {
                    var a = 1;
                }
                while (DateTime.Now.Second % 2 == 0)
                {
                    myString += ""a"";
                }".WrapInCSharpMethod();
            var fixtest = @"
                var myString = """";
                while (DateTime.Now.Second % 2 == 1)
                {
                    var a = 1;
                }
                var builder = new System.Text.StringBuilder();
                builder.Append(myString);
                while (DateTime.Now.Second % 2 == 0)
                {
                    builder.Append(""a"");
                }
                myString = builder.ToString();".WrapInCSharpMethod();
            await VerifyCSharpFixAsync(source, fixtest);
        }

        [Fact]
        public async Task FixesAddAssignmentWithoutClashingTheBuilderName()
        {
            var source = @"
                var builder = 1;
                var myString = """";
                while (DateTime.Now.Second % 2 == 0)
                {
                    myString += ""a"";
                }".WrapInCSharpMethod();
            var fixtest = @"
                var builder = 1;
                var myString = """";
                var builder1 = new System.Text.StringBuilder();
                builder1.Append(myString);
                while (DateTime.Now.Second % 2 == 0)
                {
                    builder1.Append(""a"");
                }
                myString = builder1.ToString();".WrapInCSharpMethod();
            await VerifyCSharpFixAsync(source, fixtest);
        }

        [Fact]
        public async Task FixesAddAssignmentWithoutClashingTheBuilderNameOnAField()
        {
            const string source = @"
    namespace ConsoleApplication1
    {
        class TypeName
        {
            private int builder1;
            public void Foo()
            {
                var builder = 1;
                var myString = """";
                while (DateTime.Now.Second % 2 == 0)
                {
                    myString += ""a"";
                }
            }
        }
    }";
            const string fixtest = @"
    namespace ConsoleApplication1
    {
        class TypeName
        {
            private int builder1;
            public void Foo()
            {
                var builder = 1;
                var myString = """";
                var builder2 = new System.Text.StringBuilder();
                builder2.Append(myString);
                while (DateTime.Now.Second % 2 == 0)
                {
                    builder2.Append(""a"");
                }
                myString = builder2.ToString();
            }
        }
    }";
            await VerifyCSharpFixAsync(source, fixtest);
        }

        [Fact]
        public async Task ForWithStringConcatOnLocalVariableCreatesDiagnostic()
        {
            var source = @"
                var myString = """";
                for (;;)
                {
                    myString += """";
                }".WrapInCSharpMethod();
            var expected = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(14, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "myString"));
            await VerifyCSharpDiagnosticAsync(source, expected);
        }

        [Fact]
        public async Task ForWithStringConcatOnLoopLocalVariableShouldNotCreateDiagnostic()
        {
            var source = @"
                for (;;)
                {
                    var myString = """";
                    myString += """";
                    Console.WriteLine(myString);
                }".WrapInCSharpMethod();

            await VerifyCSharpHasNoDiagnosticsAsync(source);
        }

        [Fact]
        public async Task FixesAddAssignmentInFor()
        {
            var source = @"
                var myString = """";
                for (;;)
                {
                    myString += ""a"";
                    break;
                }".WrapInCSharpMethod();
            var fixtest = @"
                var myString = """";
                var builder = new System.Text.StringBuilder();
                builder.Append(myString);
                for (;;)
                {
                    builder.Append(""a"");
                    break;
                }
                myString = builder.ToString();".WrapInCSharpMethod();
            await VerifyCSharpFixAsync(source, fixtest);
        }

        [Fact]
        public async Task ForeachWithtStringConcatOnLocalVariableCreatesDiagnostic()
        {
            var source = @"
                var myString = """";
                foreach (var i in new [] {1,2,3})
                {
                    myString += """";
                }".WrapInCSharpMethod();
            var expected = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(14, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "myString"));
            await VerifyCSharpDiagnosticAsync(source, expected);
        }

        [Fact]
        public async Task FixesAddAssignmentInForeach()
        {
            var source = @"
                var myString = """";
                foreach (var i in new [] {1,2,3})
                {
                    myString += ""a"";
                }".WrapInCSharpMethod();
            var fixtest = @"
                var myString = """";
                var builder = new System.Text.StringBuilder();
                builder.Append(myString);
                foreach (var i in new [] {1,2,3})
                {
                    builder.Append(""a"");
                }
                myString = builder.ToString();".WrapInCSharpMethod();
            await VerifyCSharpFixAsync(source, fixtest);
        }

        [Fact]
        public async Task DoWithtStringConcatOnLocalVariableCreatesDiagnostic()
        {
            var source = @"
                var myString = """";
                do
                {
                    myString += """";
                } while (DateTime.Now.Second % 2 == 0);".WrapInCSharpMethod();
            var expected = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(14, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "myString"));
            await VerifyCSharpDiagnosticAsync(source, expected);
        }

        [Fact]
        public async Task FixesAddAssignmentInDo()
        {
            var source = @"
                var myString = """";
                do
                {
                    myString += ""a"";
                } while (DateTime.Now.Second % 2 == 0);".WrapInCSharpMethod();
            var fixtest = @"
                var myString = """";
                var builder = new System.Text.StringBuilder();
                builder.Append(myString);
                do
                {
                    builder.Append(""a"");
                } while (DateTime.Now.Second % 2 == 0);
                myString = builder.ToString();".WrapInCSharpMethod();
            await VerifyCSharpFixAsync(source, fixtest);
        }

        [Fact]
        public async Task WhileWithStringConcatOnFieldWithAccessCreatesDiagnostic()
        {
            const string source = @"
    using System;
    namespace ConsoleApplication1
    {
        class SomeObject
        {
			public string A = "";
        }
        class TypeName
        {
            private SomeObject someObject = new SomeObject();
            public void Foo()
            {
                while (DateTime.Now.Second % 2 == 0)
                {
                    someObject.A += """";
                }
            }
        }
    }";
            var expected = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(16, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "someObject.A"));
            await VerifyCSharpDiagnosticAsync(source, expected);
        }

        [Fact]
        public async Task WhileWithStringConcatOnFieldWithAccessOnArrayCreatesDiagnostic()
        {
            const string source = @"
    using System;
    namespace ConsoleApplication1
    {
        class SomeObject
        {
			public string[] A;
        }
        class TypeName
        {
            private SomeObject someObject = new SomeObject();
            public void Foo()
            {
                while (DateTime.Now.Second % 2 == 0)
                {
                    someObject.A[DateTime.Now.Second] += """";
                }
            }
        }
    }";
            var expected = new DiagnosticResult(DiagnosticId.StringBuilderInLoop.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(16, 21)
                .WithMessage(string.Format(StringBuilderInLoopAnalyzer.MessageFormat, "someObject.A[DateTime.Now.Second]"));
            await VerifyCSharpDiagnosticAsync(source, expected);
        }

        [Fact]
        public async Task ForWithAssignmentOnIntArrayDoesNotCreateDiagnostic()
        {
            const string source = @"
    using System;
    namespace ConsoleApplication1
    {
        class TypeName
        {
            public void Foo()
            {
				var length = 12;
				var baseOffset = 4;
				var branches = new int[length];
				for (int i = 0; i < length; i++)
					branches[i] = baseOffset + int.Parse("""");
            }
        }
    }";
            await VerifyCSharpHasNoDiagnosticsAsync(source);
        }

        [Fact]
        public async Task WhileWithoutStringConcatWithMethodParameterDoesNotCreateDiagnostic()
        {
            var source = @"
public void Looper(int a = 0)
{
    while (a < 10)
    {
        a += 1;
    }
}".WrapInCSharpClass();
            await VerifyCSharpHasNoDiagnosticsAsync(source);
        }

        [Fact]
        public async Task WhileWithoutStringConcatWithMethodRefParameterDoesNotCreateDiagnostic()
        {
            var source = @"
public void Looper(ref int a)
{
    a = 0;
    while (a < 10)
    {
        a += 1;
    }
}".WrapInCSharpClass();
            await VerifyCSharpHasNoDiagnosticsAsync(source);
        }

        [Fact]
        public async Task IgnoreWhenVariableInTheLoopContextIsChanged()
        {
            const string source = @"
class MyObject
{
    public string MyObjectString;
}
class MyClass
{
    private readonly System.Collections.Generic.List<MyObject> items = new System.Collections.Generic.List<MyObject>();
    private void M(string suffix)
    {
        foreach (MyObject o in items)
        {
            o.MyObjectString += suffix;
        }
    }
}
    }";
            await VerifyCSharpHasNoDiagnosticsAsync(source);
        }
    }
}

File: test/VisualBasic/CodeCracker.Test/Performance/StringBuilderInLoopTests.vb
Imports CodeCracker.VisualBasic.Performance
Imports Microsoft.CodeAnalysis.Testing
Imports Xunit

Namespace Performance
    Public Class StringBuilderInLoopTests
        Inherits CodeFixVerifier(Of StringBuilderInLoopAnalyzer, StringBuilderInLoopCodeFixProvider)

        <Fact>
        Public Async Function WhileWithoutAddAssignmentExpressionDoesNotCreateDiagnostic() As Task
            Const source = "
Namespace ConsoleApplication1
    Class TypeName
        Public Sub Foo()
            While (DateTime.Now.Second Mod 2 = 0)
                Method()
            End While
        End Sub
        Public Sub Method()

        End Sub
    End Class
End Namespace"
            Await VerifyBasicHasNoDiagnosticsAsync(source)
        End Function

        <Fact>
        Public Async Function WhileWithoutStringConcatDoesNotCreateDiagnostic() As Task
            Dim source = "
            Dim a = 0
            While A < 10
                a += 1
            End While".WrapInVBMethod()

            Await VerifyBasicHasNoDiagnosticsAsync(source)
        End Function

        <Fact>
        Public Async Function WhileWithoutStringConcatWithMethoParameterDoesNotCreateDiagnostic() As Task
            Const source = "
    Public Class TypeName
        Public Sub Looper(ByRef a As Integer)
            While a < 10
                a += 1
            End While
        End Sub
    End Class"

            Await VerifyBasicHasNoDiagnosticsAsync(source)
        End Function

        <Fact>
        Public Async Function WhileWithStringConcatOnLocalVariableCreateDiagnostic() As Task
            Dim source = "
            Dim a = """"
            While DateTime.Now.Second mod 2 = 10
                a += """"
            End While".WrapInVBMethod()

            Dim expected = GetExpected()

            Await VerifyBasicDiagnosticAsync(source, expected)
        End Function

        <Fact>
        Public Async Function WhileWithStringConcatOnFieldVariableCreatesDiagnostic() As Task
            Const source = "
Namespace ConsoleApplication1
    Class TypeName
        Private a As String = """"
        Public Sub Foo()
            While (DateTime.Now.Second Mod 2 = 0)
                a += """"
            End While
        End Sub
    End Class
End Namespace"

            Dim expected = New DiagnosticResult(StringBuilderInLoopAnalyzer.Id, Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(7, 17) _
                .WithMessage(String.Format(StringBuilderInLoopAnalyzer.MessageFormat, "a"))
            Await VerifyBasicDiagnosticAsync(source, expected)
        End Function

        Private Shared Function GetExpected() As DiagnosticResult
            Return New DiagnosticResult(StringBuilderInLoopAnalyzer.Id, Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(9, 17) _
                .WithMessage(String.Format(StringBuilderInLoopAnalyzer.MessageFormat, "a"))
        End Function

        <Fact>
        Public Async Function WhileWithStringConcatOnLocalVariableCreatesDiagnostic() As Task
            Dim source = "
            Dim a = """"
            While DateTime.Now.Second mod 2 = 0
                a += """"
            End While
".WrapInVBMethod()

            Await VerifyBasicDiagnosticAsync(source, GetExpected())
        End Function

        <Fact>
        Public Async Function WhileWithStringConcatOnPropertyVariableCreatesDiagnostic() As Task
            Const source = "
Namespace ConsoleApplication1
    Class TypeName
        Private Property a As String = """"
        Public Sub Foo()
            While (DateTime.Now.Second Mod 2 = 0)
                a += """"
            End While
        End Sub
    End Class
End Namespace"

            Dim expected = New DiagnosticResult(StringBuilderInLoopAnalyzer.Id, Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(7, 17) _
                .WithMessage(String.Format(StringBuilderInLoopAnalyzer.MessageFormat, "a"))

            Await VerifyBasicDiagnosticAsync(source, expected)
        End Function
        <Fact>
        Public Async Function WhileWithStringConcatWithSeveralConcatsOnDifferentVarsCreatesSeveralDiagnostics() As Task
            Dim source = "
            Dim a = """"
            Dim myString2 = """"
            While DateTime.Now.Second mod 2 = 0
                a += """"
                myString2 += """"
            End While
            Console.WriteLine(myString2)
".WrapInVBMethod()

            Dim expected1 = New DiagnosticResult(StringBuilderInLoopAnalyzer.Id, Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(10, 17) _
                .WithMessage(String.Format(StringBuilderInLoopAnalyzer.MessageFormat, "a"))
            Dim expected2 = New DiagnosticResult(StringBuilderInLoopAnalyzer.Id, Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(11, 17) _
                .WithMessage(String.Format(StringBuilderInLoopAnalyzer.MessageFormat, "myString2"))

            Await VerifyBasicDiagnosticAsync(source, expected1, expected2)
        End Function

        <Fact>
        Public Async Function WhileWithStringConcatWithSimpleAssignmentCreatesDiagnostic() As Task
            Dim source = "
            Dim a = """"
            While DateTime.Now.Second mod 2 = 0
                a = a + """"
            End While
".WrapInVBMethod()

            Await VerifyBasicDiagnosticAsync(source, GetExpected())
        End Function

        <Fact>
        Public Async Function WhileWithStringConcatWithSimpleAssignmentOnDifferentDimDoesNotCreateDiagnostic() As Task
            Dim source = "Dim a = """"
            Dim otherString = """"
            While DateTime.Now.Second Mod 2 = 0
                a = otherString + """"
            End While
".WrapInVBMethod
            Await VerifyBasicHasNoDiagnosticsAsync(source)
        End Function

        <Fact>
        Public Async Function FixesAddAssignmentInWhile() As Task
            Dim source = "Dim a = """"
            While DateTime.Now.Second Mod 2 = 0
                a += ""a""
            End While
".WrapInVBMethod

            Dim fix = "Dim a = """"
            Dim builder As New Text.StringBuilder()
            builder.Append(a)
            While DateTime.Now.Second Mod 2 = 0
                builder.Append(""a"")
            End While
            a = builder.ToString()
".WrapInVBMethod()
            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function FixesAddAssignmentInWhileWithSystemTextInContext() As Task
            Const source = "
Imports System
Imports System.Text
Namespace ConsoleApplication1

    Class TypeName
        Public Sub Test()
            Dim a = """"
            While (DateTime.Now.Second Mod 2 = 0)
                a += ""a""
            End While
        End Sub
    End Class
End Namespace"


            Const fix = "
Imports System
Imports System.Text
Namespace ConsoleApplication1

    Class TypeName
        Public Sub Test()
            Dim a = """"
            Dim builder As New StringBuilder()
            builder.Append(a)
            While (DateTime.Now.Second Mod 2 = 0)
                builder.Append(""a"")
            End While
            a = builder.ToString()
        End Sub
    End Class
End Namespace"

            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function FixesSimpleAssignmentInWhile() As Task
            Dim source = "Dim a = """"
            ' comment 3
            While (DateTime.Now.Second Mod 2 = 0)
                ' comment 1
                a += ""a"" 'comment 2
            End While 'comment 4
".WrapInVBMethod

            Dim fix = "Dim a = """"
            Dim builder As New Text.StringBuilder()
            builder.Append(a)
            ' comment 3
            While (DateTime.Now.Second Mod 2 = 0)
                ' comment 1
                builder.Append(""a"") 'comment 2
            End While 'comment 4
            a = builder.ToString()
".WrapInVBMethod

            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function FixesAddAssignmentWhenThereAre2WhilesOnBlock() As Task
            Dim source = "Dim a = """"
            While (DateTime.Now.Second Mod 2 = 0)
                Dim a = 1
            End While
            While (DateTime.Now.Second Mod 2 = 0)
                a += ""a""
            End While
".WrapInVBMethod()

            Dim fix = "Dim a = """"
            While (DateTime.Now.Second Mod 2 = 0)
                Dim a = 1
            End While
            Dim builder As New Text.StringBuilder()
            builder.Append(a)
            While (DateTime.Now.Second Mod 2 = 0)
                builder.Append(""a"")
            End While
            a = builder.ToString()
".WrapInVBMethod()

            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function FixesAddAssignmentWithoutClashingTheBuilderName() As Task
            Dim source = "Dim builder = 1
            Dim a = """"
            While (DateTime.Now.Second Mod 2 = 0)
                a += ""a""
            End While
".WrapInVBMethod()

            Dim fix = "Dim builder = 1
            Dim a = """"
            Dim builder1 As New Text.StringBuilder()
            builder1.Append(a)
            While (DateTime.Now.Second Mod 2 = 0)
                builder1.Append(""a"")
            End While
            a = builder1.ToString()
".WrapInVBMethod()

            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function FixesAddAssignmentWithoutClashingTheBuilderNameOnAField() As Task
            Const source = "
Namespace ConsoleApplication1

    Class TypeName
        Private builder As Integer
        Public Sub Foo()
            Dim builder = 1
            Dim a = """"
            While (DateTime.Now.Second Mod 2 = 0)
                a += ""a""
            End While
        End Sub
    End Class
End Namespace"


            Const fix = "
Namespace ConsoleApplication1

    Class TypeName
        Private builder As Integer
        Public Sub Foo()
            Dim builder = 1
            Dim a = """"
            Dim builder1 As New System.Text.StringBuilder()
            builder1.Append(a)
            While (DateTime.Now.Second Mod 2 = 0)
                builder1.Append(""a"")
            End While
            a = builder1.ToString()
        End Sub
    End Class
End Namespace"

            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function ForWithStringConcatOnLocalVariableCreatesDiagnostic() As Task
            Dim source = "Dim a = """"
            For i As Integer = 1 To 10
                a += ""a""
            Next".WrapInVBMethod

            Dim fix = "Dim a = """"
            Dim builder As New Text.StringBuilder()
            builder.Append(a)
            For i As Integer = 1 To 10
                builder.Append(""a"")
            Next
            a = builder.ToString()".WrapInVBMethod

            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function FixesAddAssignmentInFor() As Task
            Dim source = "Dim a = """"
            For i As Integer = 1 To 10
                a += ""b""
                Exit For
            Next".WrapInVBMethod

            Dim builder As New System.Text.StringBuilder()
            builder.Append("a")

            Dim fix = "Dim a = """"
            Dim builder As New Text.StringBuilder()
            builder.Append(a)
            For i As Integer = 1 To 10
                builder.Append(""b"")
                Exit For
            Next
            a = builder.ToString()".WrapInVBMethod

            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function ForeachWithStringConcatOnLocalVariableCreatesDiagnostic() As Task
            Dim source = "
            Dim a = """"
            For Each i In {1, 2, 3}
                a += """"
            Next".WrapInVBMethod

            Dim expected = New DiagnosticResult(StringBuilderInLoopAnalyzer.Id, Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(9, 17) _
                .WithMessage(String.Format(StringBuilderInLoopAnalyzer.MessageFormat, "a"))
            Await VerifyBasicDiagnosticAsync(source, expected)

        End Function

        <Fact>
        Public Async Function FixesAddAssignmentInForEach() As Task
            Dim source = "Dim a = """"
            For Each i In {1, 2, 3}
                a += ""a""
            Next".WrapInVBMethod

            Dim fix = "Dim a = """"
            Dim builder As New Text.StringBuilder()
            builder.Append(a)
            For Each i In {1, 2, 3}
                builder.Append(""a"")
            Next
            a = builder.ToString()".WrapInVBMethod

            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function DoWithStringConcatOnLocalVariableCreatesDiagnostic() As Task
            Dim source = "
            Dim a = """"
            Do
                a += """"
            Loop Until DateTime.Now.Second Mod 2 = 0
".WrapInVBMethod

            Dim expected = New DiagnosticResult(StringBuilderInLoopAnalyzer.Id, Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(9, 17) _
                .WithMessage(String.Format(StringBuilderInLoopAnalyzer.MessageFormat, "a"))
            Await VerifyBasicDiagnosticAsync(source, expected)
        End Function

        <Fact>
        Public Async Function FixesAddAssignmentInForDo() As Task
            Dim source = "Dim a = """"
            Do
                a += ""a""
            Loop Until DateTime.Now.Second Mod 2 = 0
".WrapInVBMethod

            Dim b As New System.Text.StringBuilder()

            Dim fix = "Dim a = """"
            Dim builder As New Text.StringBuilder()
            builder.Append(a)
            Do
                builder.Append(""a"")
            Loop Until DateTime.Now.Second Mod 2 = 0
            a = builder.ToString()
".WrapInVBMethod

            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function ForLoopInMethodWithoutStringShouldNotCreateDiagnostic() As Task
            Const source = "
    Public Class Test
        Private Sub AdjustSample(ByRef readIndex As Integer, writeBuffer() As Single, ByRef writeIndex As Integer)
            For i = 0 To 2
                writeBuffer(writeIndex) = 0
                readIndex += 1
                writeIndex += 1
            Next
        End Sub
    End Class"

            Await VerifyBasicHasNoDiagnosticsAsync(source)
        End Function

        <Fact>
        Public Async Function ForLoopStringDeclaredAndConcatenatedWithinShouldNotCreateDiagnostic() As Task
            Dim source = "
            For i = 0 To 10
                Dim someString As String = i.ToString()
                someString &= "" ""
                Console.WriteLine(someString)
            Next".WrapInVBMethod()

            Await VerifyBasicHasNoDiagnosticsAsync(source)
        End Function
    End Class
End Namespace